File: Coding/240/searchMatrix.py
class Solution(object):
    # Time efficiency: O(n*m), Space Complexity: O(1)
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        cur = 0
        
        for i in range(len(matrix)):
            increase = True
            if cur < 0:
                return False
            cur = min(len(matrix[0])-1,cur)
            
            if matrix[i][cur] > target:
                increase = False
            
            while 0 <= cur < len(matrix[0]):
                if matrix[i][cur]==target:
                    return True
                if increase:
                    if matrix[i][cur] > target:
                        break
                    else:
                        cur+=1
                else:
                    if matrix[i][cur] < target:
                        break
                    else:
                        cur-=1
        return False

    # Time efficiency: O(m+n), Space Complexity: O(1)
    # When we try to find the target, first observe the relationship between each point.
    def searchMatrix2(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        cur = [0,len(matrix[0])-1]
        
        while cur[0] < len(matrix) and cur[1] >= 0:
            if matrix[cur[0]][cur[1]]==target:
                return True
            elif matrix[cur[0]][cur[1]]>target:
                cur[1]-=1
            else:
                cur[0]+=1
        return False

File: Coding/240/searchMatrix.cpp
class Solution {
public:
    // Time efficiency: O(m+n), Space efficiency: O(1)
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        int n = matrix[0].size();
        
        int i = 0;
        int j = matrix[0].size()-1;
        
        while (i<m && j>=0){
            if (matrix[i][j] == target){
                return true;
            } else if (matrix[i][j] < target){
                i+=1;
            } else {
                j-=1;
            }
        }
        return false;
    }

    // Time efficiency: O(mlogn), Space efficiency: O(1)
    bool searchMatrix2(vector<vector<int>>& matrix, int target) {
        for(int i = 0; i < matrix.size(); i++){
            int left = 0;
            int right = matrix[0].size()-1;
            while (left <= right) {
                int mid = (right+left)/2;
                if (matrix[i][mid]==target){
                    return true;
                } else if (matrix[i][mid] > target){
                    right = mid-1;
                } else {
                    left = mid+1;
                }
            }
        }
        return false;
    }
};
